iceoryx_hoofs/test/moduletests/test_static_lifetime_guard.cpp

```diff
@@ -0,0 +1,318 @@
+{
+    ::testing::Test::RecordProperty("TEST_ID", "32a2fdbf-cb02-408c-99a3-373aa66b2764");
+    INIT_TEST;
+    T::initInstance();
+
+    T::Guard guard;
+    {
+        EXPECT_EQ(T::Guard::count(), 2);
+        // NOLINTJUSTIFICATION ctor and dtor side effects are tested
+        // NOLINTNEXTLINE(performance-unnecessary-copy-initialization)
+        T::Guard movedGuard(std::move(guard));
+        EXPECT_EQ(T::Guard::count(), 3);
+    }
+    EXPECT_EQ(T::Guard::count(), 2);
+
+    EXPECT_EQ(T::Foo::ctorCalled, 0);
+    EXPECT_EQ(T::Foo::dtorCalled, 0);
+}
+
+TEST_F(StaticLifetimeGuard_test, assignmentDoesNotChangeLifetimeCount)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "1c04ac75-d47a-44da-b8dc-6f567a53d3fc");
+    INIT_TEST;
+    T::initInstance();
+
+    T::Guard guard1;
+    T::Guard guard2;
+
+    EXPECT_EQ(T::Guard::count(), 3);
+    guard1 = guard2;
+    EXPECT_EQ(T::Guard::count(), 3);
+    guard1 = std::move(guard2);
+    EXPECT_EQ(T::Guard::count(), 3);
+
+    EXPECT_EQ(T::Foo::ctorCalled, 0);
+    EXPECT_EQ(T::Foo::dtorCalled, 0);
+}
+
+TEST_F(StaticLifetimeGuard_test, destructionAtZeroCountWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "8b5a22a9-87bc-434b-9d07-9f3c20a6944e");
+    INIT_TEST;
+    T::initInstance();
+
+    {
+        T::Guard guard;
+        auto& instance = T::Guard::instance();
+
+        // count is expected to be 2,
+        // we ignore the guard of testInstance() by setting it to 1,
+        // hence when guard is destroyed the instance will be destroyed as well
+        auto oldCount = T::setCount(1);
+        EXPECT_EQ(oldCount, 2);
+
+        EXPECT_EQ(T::Foo::ctorCalled, 0);
+        EXPECT_EQ(T::Foo::dtorCalled, 0);
+        EXPECT_EQ(instance.id, FIRST_INSTANCE_ID);
+    }
+
+    EXPECT_EQ(T::Guard::count(), 0);
+    EXPECT_EQ(T::Foo::ctorCalled, 0);
+    EXPECT_EQ(T::Foo::dtorCalled, 1);
+}
+
+TEST_F(StaticLifetimeGuard_test, constructionAfterDestructionWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "0077e73d-ddf5-47e7-a7c6-93819f376175");
+    INIT_TEST;
+    T::initInstance();
+
+    {
+        T::Guard guard;
+        auto& instance = T::Guard::instance();
+
+        T::setCount(1);
+        EXPECT_EQ(instance.id, FIRST_INSTANCE_ID);
+    }
+
+    // first instance destroyed (should usually only happen at the the program
+    // during static destruction)
+
+    T::Foo::reset();
+
+    EXPECT_EQ(T::Guard::count(), 0);
+    {
+        T::Guard guard;
+        auto& instance = T::Guard::instance();
+
+        EXPECT_EQ(T::Foo::ctorCalled, 1);
+        EXPECT_EQ(T::Foo::dtorCalled, 0);
+        EXPECT_EQ(instance.id, SECOND_INSTANCE_ID);
+    }
+
+    // there was only one guard for the second instance that is destroyed
+    // at scope end and hence the second instance should be destroyed as well
+
+    EXPECT_EQ(T::Guard::count(), 0);
+    EXPECT_EQ(T::Foo::ctorCalled, 1);
+    EXPECT_EQ(T::Foo::dtorCalled, 1);
+}
+
```

Comment:
Does it make sense to add a test where we check concurrently that the c'tor `T()` is only called once in `StaticLifetimeGuard<T>::instance(Args&&... args)`?

Reply:
It does. The test will just be a little complicated/not 100% accurate due to by nature depending on thread timing for the interesting case (where construction is attempted separately).

I will come up with something. Concurrency is hard but interesting and our only hope to scale computation so we need to sharpen our skills in using it correctly :-)

Reply:
Added a concurrent test. Note that this cannot guarantee the required execution scenario as the threads may e.g. run sequentially. This is unavoidable with the thread scheduling we can assume.